src/lower/merge_lattice.cpp

```diff
@@ -577,10 +619,49 @@ class MergeLatticeBuilder : public IndexNotationVisitorStrict, public IterationA
   {
     vector<MergePoint> points;
 
+    struct pointSort {
+      bool operator()(const MergePoint& a, const MergePoint& b) {
+        size_t left_size  = a.iterators().size() + a.locators().size();
+        size_t right_size = b.iterators().size() + b.locators().size();
+        return left_size > right_size;
+      }
+    } pointSorter;
+
     // Append all combinations of the merge points of a and b
-    for (auto& apoint : left.points()) {
-      for (auto& bpoint : right.points()) {
-        points.push_back(unionPoints(apoint, bpoint));
+    auto sorted_apoint = left.points();
+    auto sorted_bpoint = right.points();
+    std::sort(sorted_apoint.begin(), sorted_apoint.end(), pointSorter);
+    std::sort(sorted_bpoint.begin(), sorted_bpoint.end(), pointSorter);
+
+    set<Iterator> apoint_root_set;
+    if (!sorted_apoint.empty())
+      apoint_root_set = sorted_apoint.begin()->tensorRegion();
+
+    set<Iterator>bpoint_root_set;
+    if (!sorted_bpoint.empty())
+      bpoint_root_set = sorted_bpoint.begin()->tensorRegion();
+
+    for (auto& apoint : sorted_apoint) {
+      for (auto& bpoint : sorted_bpoint) {
+        bool hasIntersection = true;
+
+        auto apoint_set = apoint.tensorRegion();
+        auto bpoint_set = bpoint.tensorRegion();
+
+        for (auto& it : apoint_set) {
+          if (!std::count(bpoint_set.begin(), bpoint_set.end(), it) &&
```

Comment:
Same question as above about using util::contains

Reply:
Will change this and push directly to array_algebra when I have time. Same as above.